### PR TITLE
fix exception handling of downloads in new-build

### LINKED ISSUE #3387: confusing error-message "thread blocked indefinitely in an MVar operation" when downloading tarball fails
when download a source-tarball fails due to missing file we currently get a somewhat unfortunate error message:

```
$ cabal new-build -j1 exe:cabal -v3
File monitor 'improved-plan' unchanged.
In order, the following will be built:
HTTP-4000.3.3 (lib:HTTP) (requires download & build)
cabal-install-1.24.0.0 (exe:cabal) (configuration changed)
creating /stuff3/work/GitHub/cabal/dist-newstyle/build
creating /stuff3/work/GitHub/cabal/dist-newstyle/tmp
Downloading HTTP-4000.3.3...
writing
/home/hvr/.cabal/packages/hackage.haskell.org/HTTP/4000.3.3/HTTP-4000.3.3.tar.gz
Ready to install HTTP-4000.3.3
Selected mirror http://hackage.haskell.org/
Waiting for install task to finish...
Waiting for download of HTTP-4000.3.3 to finish
Downloading package HTTP-4000.3.3
Sending:
GET /package/HTTP-4000.3.3.tar.gz HTTP/1.1
Host: hackage.haskell.org
User-Agent: cabal-install/1.24.0.0 (linux; x86_64)
Cache-Control: no-transform

Creating new connection to hackage.haskell.org
Received:
HTTP/1.1 404 Not Found
Server: nginx/1.10.0 (Ubuntu)
Date: Mon, 02 May 2016 09:23:40 GMT
Content-Type: text/html
Content-Length: 178
Connection: keep-alive

Exception Unexpected response 404for
http://hackage.haskell.org/package/HTTP-4000.3.3.tar.gz when using mirror
http://hackage.haskell.org/
Selected mirror http://ftp.osuosl.org/pub/hackage
Downloading package HTTP-4000.3.3
Sending:
GET /pub/hackage/package/HTTP-4000.3.3.tar.gz HTTP/1.1
Host: ftp.osuosl.org
User-Agent: cabal-install/1.24.0.0 (linux; x86_64)
Cache-Control: no-transform

Creating new connection to ftp.osuosl.org
Received:
HTTP/1.1 404 Not Found
Date: Mon, 02 May 2016 09:24:35 GMT
Server: Apache
Content-Length: 302
Content-Type: text/html; charset=iso-8859-1

thread blocked indefinitely in an MVar operation
```

/cc @dcoutts

### PR DESCRIPTION
**Pull request by dcoutts**
Partly this is a continuation of the InstallPlan refactoring #3622, but also a related change to fix #3387, by handling exceptions during download properly.

These things are also a step towards finishing #3394 properly, since it means we now have all the `BuildResults` using proper exception types and so will be able to use that to report at the end what if anything when wrong.

### CODE REVIEW

File: cabal-install/Distribution/Client/FetchUtils.hs [hunk 1]
```diff
@@ -186,6 +197,61 @@ downloadIndex transport verbosity remoteRepo cacheDir = do
...
+waitAsyncFetchPackage :: Verbosity
+                      -> AsyncFetchMap
+                      -> UnresolvedPkgLoc
+                      -> IO ResolvedPkgLoc
+waitAsyncFetchPackage verbosity downloadMap srcloc =
+    case Map.lookup srcloc downloadMap of
+      Just hnd -> do
+        debug verbosity $ "Waiting for download of " ++ show srcloc
+        either throwIO return =<< takeMVar hnd
+      Nothing -> fail "waitAsyncFetchPackage: package not being download"
```
Comment (ezyang) [inline cabal-install/Distribution/Client/FetchUtils.hs]: downloaded

File: cabal-install/Distribution/Client/FetchUtils.hs [hunk 2]
```diff
@@ -186,6 +197,61 @@ downloadIndex transport verbosity remoteRepo cacheDir = do
...
+-- | Fork off an async action to download the given packages (by location).
+--
+-- The downloads are initiated in order, so you can arrange for packages that
+-- will likely be needed sooner to be earlier in the list.
+--
+-- The body action is passed a map from those packages (identified by their
+-- location) to a completion var for that package. So the body action should
+-- lookup the location and use 'asyncFetchPackage' to get the result.
+--
+asyncFetchPackages :: Verbosity
```
Comment (ezyang) [inline cabal-install/Distribution/Client/FetchUtils.hs]: If you're using the async package, why not just actually use the combinators on `Async` to implement this? Using `withAsync` and then ignoring the passed in `Async` just seems extremely strange.

**Comment by ezyang** [comment]
I don't really understand the patch, but if it fixes the exceptions problem :+1:

File: cabal-install/Distribution/Client/FetchUtils.hs (hunk 2 above)
Comment (dcoutts) [inline cabal-install/Distribution/Client/FetchUtils.hs]: What I used to do was fork them all off using async and put the async handles in the map. The problem with that is we don't actually wan to do them all concurrently, and also we want to do them in a particular order, to get the ones we need back soonest. I don't see any way to implement this while still fitting the async API.

**Comment by dcoutts** [comment]
> I don't really understand the patch, but if it fixes the exceptions problem :+1:

Which one, the first one or the downloads one?

File: cabal-install/Distribution/Client/FetchUtils.hs [hunk 3]
```diff
@@ -186,6 +197,61 @@ downloadIndex transport verbosity remoteRepo cacheDir = do
...
+
+    asyncDownloadVars <- sequence [ do v <- newEmptyMVar
+                                       return (pkgloc, v)
+                                  | pkgloc <- pkglocs ]
+
+    let fetchPackages :: IO ()
+        fetchPackages =
+          forM_ asyncDownloadVars $ \(pkgloc, var) -> do
+            result <- try $ fetchPackage verbosity repoCtxt pkgloc
+            putMVar var result
```
Comment (dcoutts) [inline cabal-install/Distribution/Client/FetchUtils.hs]: This is the crucial bit really, for fixing exception handling. Previously the exception was propagated locally in the async action and so the callers picking up the async fetch package results would wait forever (or rather get an MVar deadlock exception).

File: cabal-install/Distribution/Client/FetchUtils.hs [hunk 4]
```diff
@@ -186,6 +197,61 @@ downloadIndex transport verbosity remoteRepo cacheDir = do
...
+--
+waitAsyncFetchPackage :: Verbosity
+                      -> AsyncFetchMap
+                      -> UnresolvedPkgLoc
+                      -> IO ResolvedPkgLoc
+waitAsyncFetchPackage verbosity downloadMap srcloc =
+    case Map.lookup srcloc downloadMap of
+      Just hnd -> do
+        debug verbosity $ "Waiting for download of " ++ show srcloc
+        either throwIO return =<< takeMVar hnd
```
Comment (dcoutts) [inline cabal-install/Distribution/Client/FetchUtils.hs]: Plus this bit to rethrow.

**Comment by ezyang** [comment]
My lack of understanding is mostly cdacc518807e479855abc4981c1871a8fc651cec; there's some shuffling around of types, but I don't really know how things were structured previously, or how they are structured now. But... code got removed, so I'm happy to take your word for it ;)

**Comment by dcoutts** [comment]
So the new-build code previously used:
```
data GenericBuildResult ipkg iresult ifailure
                  = BuildFailure ifailure
                  | BuildSuccess [ipkg] iresult
type BuildResult  = GenericBuildResult InstalledPackageInfo
                                       BuildSuccess BuildFailure
data BuildSuccess = BuildOk DocsResult TestsResult
data BuildFailure = ...
```
whereas the common D.C.Types used by the D.C.Install etc code uses:
```
type BuildResult  = Either BuildFailure BuildSuccess
 data BuildSuccess = BuildOk DocsResult TestsResult [InstalledPackageInfo]
data BuildFailure = ...
```
So the main difference (apart from using Either vs a custom 2-constructor type) is that the previous new-build code put the `[InstalledPackageInfo]` into the `BuildResult` constructor separately from the `BuildSuccess` type. So that's where the main shuffling is happening. Otherwise it's just a matter of using the types defined in D.C.Types.

**Comment by dcoutts** [comment]
I don't understand the travis failure on 7.4.2. It looks like it's cabal sdist that is failing, but the code in this PR is completely unrelated to sdist.

**Comment by ezyang** [comment]
I think it's OK if we disable sdist test in the main dist; I've found that it doesn't work very reliably if there is configuration (as there is here) for a differing version of Cabal; if you clean it works. We do have a test for sdist in bootstrap.

**Comment by dcoutts** [comment]
Ok to merge then?

**Comment by ezyang** [comment]
Yep, but you or I will have to immediately push a commit to remove the sdist from `travis-script.sh`

**Comment by 23Skidoo** [comment]
LGTM as well.

**Pull request merged**